tempesta_fw/http.c

```diff
@@ -1089,58 +1108,130 @@ tfw_http_req_evict(TfwSrvConn *srv_conn, TfwServer *srv, TfwHttpReq *req,
+			      TfwHttpReq *req, FwdState *st)
+{
+	int budget = 10;
+	int ret = -1;
+	unsigned short prev_state = st->state;
+
+	BUG_ON(st->state == FWD_STATE_RESCH);
+	__http_req_delist(conn, req);
+	/*
+	 * Explicitly mark current connection as busy to exclude it from
+	 * normal rescheduling and, as a consiquence, to avoid deadlock
+	 * on it during normal rescheduling.
+	 *
+	 * NOTE: Connection will be unmarked after all requests in current
+	 * connection will be processed: in tfw_http_conn_fwd_unsent() or
+	 * in tfw_http_conn_resend().
+	 */
+	tfw_srv_conn_set_busy(conn);
+	st->state = FWD_STATE_RESCH;
+
+	while (budget--) {
+		BUG_ON(st->state != FWD_STATE_RESCH);
+		if (tfw_http_req_evict_retries(NULL, srv, req, &st->eq)
+		    || tfw_http_req_resched(req, srv, st)
+		    || st->state == FWD_STATE_RESCH_FIN)
+		{
+			ret = 0;
+			break;
+		}
+	}
+
+	st->state = prev_state;
+	return ret;
+}
+
+/*
+ * If forwarding of @req to server @srv_conn is not successful, requests
+ * should be left in @fwd_queue (to resend them to backend later) in
+ * following cases:
+ * - if connection is not active (-EBADF is returned);
+ * - if connection is busy and @srv_conn->msg_sent is not NULL; forwarding
+ *   process will be continued later, when connection will be reestablished
+ *   or when the response for unanswered request will arrive from backend;
+ *   this case is not applicable for re-send forwarding state;
+ * - if connection is already in reschedule forwarding state; this case is
+ *   necessary to avoid infinite recursion during rescheduling on busy
+ *   connections;
+ * If connection is busy but has not unanswered requests, then reschedule
+ * procedure takes place for it. In other cases request must be moved to
+ * the error queue @st->eq (for sending an error response later).
+ */
+static inline int
 tfw_http_req_fwd_send(TfwSrvConn *srv_conn, TfwServer *srv,
-		      TfwHttpReq *req, struct list_head *eq)
+		      TfwHttpReq *req, FwdState *st)
 {
+	int r;
+
 	req->jtxtstamp = jiffies;
 	tfw_http_req_init_ss_flags(srv_conn, req);
 
-	if (tfw_connection_send((TfwConn *)srv_conn, (TfwMsg *)req)) {
-		TFW_DBG2("%s: Forwarding error: conn=[%p] req=[%p]\n",
-			 __func__, srv_conn, req);
-		if (req->flags & TFW_HTTP_F_HMONITOR) {
-			__http_req_delist(srv_conn, req);
-			WARN_ON_ONCE(req->pair);
-			tfw_http_msg_free((TfwHttpMsg *)req);
-			TFW_WARN_ADDR("Unable to send health"
-			              " monitoring request to server",
-			              &srv_conn->peer->addr, TFW_WITH_PORT);
-		} else {
-			tfw_http_req_err(srv_conn, req, eq, 500,
-					 "request dropped: forwarding error");
-		}
-		return false;
+	if (!(r = tfw_connection_send((TfwConn *)srv_conn, (TfwMsg *)req)))
+		return 0;
+
+	if (r == -EBADF
+	    || st->state == FWD_STATE_RESCH
+	    || (r == -EBUSY
+		&& srv_conn->msg_sent != NULL
+		&& st->state != FWD_STATE_RESEND))
+		return r;
+
+	if (r == -EBUSY
+	    && !tfw_http_req_fwd_busy_resched(srv_conn, srv, req, st))
+		return r;
```

Comment:
I think this is the main problem of the patch set introducing the complexity and recursion.

I had a quick review of the rescheduling logic which we have before this change. I didn't dig deep, but I imagine following changes to make the code straightforward - maybe you can think in this direction and/or @ikoveshnikov proposal and figure out some more clean solution:

1. `tfw_http_req_fwd()` returns a status code - whether it could send a request or not, so `tfw_http_conn_fwd_unsent()` also returns a status code.
2. `tfw_http_req_resched()` also returns a status code by the result of `tfw_http_req_fwd()` call
3. move the end of `tfw_http_conn_shrink_fwdq_resched()` to a new function `good_func_name1()` which calls the request eviction and rescheduling in a loop, so we limit number of retries/reschedules.
4.  `tfw_http_req_cache_cb()` calls ``good_func_name2()` instead of `tfw_http_req_fwd()`. ``good_func_name2()` calls `tfw_http_req_fwd()` and `good_func_name1()` if the former fails.
5. `tfw_http_hm_srv_send()` just do the same as now.

Reply:
Solution had been refactored to avoid recursion and states on stack.